pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,858 @@
+	// the PacketCapture is done successfully.
+	targetCapturedPacketsNum int32
+	// updateRateLimiter controls the frequency of the updates to PacketCapture status.
+	updateRateLimiter *rate.Limiter
+	// pcapngFile is the file object for the packet file.
+	pcapngFile afero.File
+	// pcapngWriter is the writer for the packet file.
+	pcapngWriter *pcapgo.NgWriter
+}
+
+type Controller struct {
+	kubeClient            clientset.Interface
+	crdClient             clientsetversioned.Interface
+	packetCaptureInformer crdinformers.PacketCaptureInformer
+	packetCaptureLister   crdlisters.PacketCaptureLister
+	packetCaptureSynced   cache.InformerSynced
+	interfaceStore        interfacestore.InterfaceStore
+	queue                 workqueue.TypedRateLimitingInterface[string]
+	sftpUploader          ftp.Uploader
+	newPacketSourceFn     func(c *Controller, pc *crdv1alpha1.PacketCapture) (PacketSource, error)
+	cond                  *sync.Cond
+	// A name-phase mapping for all PacketCapture CRs.
+	captures           map[string]packetCapturePhase
+	numRunningCaptures int
+	waitingCaptures    int
+	waitingCapturesCh  chan string
+}
+
+func NewPacketCaptureController(
+	kubeClient clientset.Interface,
+	crdClient clientsetversioned.Interface,
+	packetCaptureInformer crdinformers.PacketCaptureInformer,
+	interfaceStore interfacestore.InterfaceStore,
+) *Controller {
+	c := &Controller{
+		kubeClient:            kubeClient,
+		crdClient:             crdClient,
+		packetCaptureInformer: packetCaptureInformer,
+		packetCaptureLister:   packetCaptureInformer.Lister(),
+		packetCaptureSynced:   packetCaptureInformer.Informer().HasSynced,
+		interfaceStore:        interfaceStore,
+		queue: workqueue.NewTypedRateLimitingQueueWithConfig(
+			workqueue.NewTypedItemExponentialFailureRateLimiter[string](minRetryDelay, maxRetryDelay),
+			workqueue.TypedRateLimitingQueueConfig[string]{Name: "packetcapture"},
+		),
+		sftpUploader:      &ftp.SftpUploader{},
+		cond:              sync.NewCond(&sync.Mutex{}),
+		captures:          make(map[string]packetCapturePhase),
+		waitingCapturesCh: make(chan string, maxWaitingCaptures),
+	}
+
+	packetCaptureInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addPacketCapture,
+		UpdateFunc: c.updatePacketCapture,
+		DeleteFunc: c.deletePacketCapture,
+	}, resyncPeriod)
+
+	c.newPacketSourceFn = func(c *Controller, pc *crdv1alpha1.PacketCapture) (PacketSource, error) {
+		matchPacket, err := c.createMatchPacket(pc)
+		if err != nil {
+			return nil, err
+		}
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil, nil
+		}
+		filter := bpf.CompilePacketFilter(pc.Spec.Packet, matchPacket)
+		klog.V(5).InfoS("Generated bpf instructions for Packetcapture", "name", pc.Name, "match packet", matchPacket, "bpf instructions", filter)
+		return NewPcapSource(*device, filter)
+	}
```

Comment:
I don't quite understand this structure. I guess the intention is to abstract how packet is captured, but the code looks strange:
1. A abstracted capturer shouldn't rely on `Controller`.
2. It doesn't make sense to involve the struct specific to openflow.
3. Isn't the construction of `BpfInstructions` specific to the implementation? Why it's kept in the controller?

I think the code could be clearer if there is an interface like the following:
```
type PacketCapturer interface {
	Capture(ctx context.Context, device, srcIP, dstIP string, packet *crdv1alpha1.Packet, num int32) (chan gopacket.Packet, error)
}
```

- The conversion from `Source` and `Destination` to `srcIP` and `dstIP` should be done in controller as it requires K8s knowledge and is reusable for all captures.
- The conversion from `device, srcIP, dstIP string, packet *crdv1alpha1.Packet` to `bpf.Instruction` should be done in the interface's implementation as it's specific to this capturer.
- There should be a `context` passed to the implementation which is used to control the blocking capture function (gopacket provides `PacketsCtx`).

Reply:
updated according suggestions.